### PR TITLE
Throw ConnectionResetException when socket closed observed by write

### PR DESCRIPTION
**Pull request by halter73**
#2546

### CODE REVIEW

**Comment by halter73** [comment]
Temporarily rebased on #2574 to remove noise from DiagnsoticMemoryPool.

**Comment by davidfowl** [comment]
Can we fix this properly all at once? Or is this just to work around the issue right now?

**Comment by halter73** [comment]
What do you mean exactly @davidfowl? None of these fixes are "workarounds". Did you read the following that I wrote in #2547? 

> The prior strategy of waiting for the pipe completed callbacks doesn't work because blocks are returned to the memory pool after the callbacks are fired.

You could argue that ConnectionClosed should be a Task instead of a CancellationToken (and you'd be right), but I figured it doesn't hurt too much to use the existing public API instead of using a new one.

**Comment by davidfowl** [comment]
I don't think the transport should be waiting on the application task. That should be in the server itself

**Comment by halter73** [comment]
@davidfowl The only reason the transport is waiting on the application task is so that it can safely dispose the CTS without risking an ODE in HttpConnectionMiddleware.OnConnectionAsync(). The only other alternatives I see are:

1. Expose the CTS on the feature and have the middleware dispose it 🤢
2. Replace the CancellationToken on IConnectionLifetimeFeature with a Task (ideal, but this breaks a now-public API)
3. Have the transports trip a CTS and a TCS because why not?

@pakrym and I knew you would hate this. Switching to a TCS means that the transport doesn't need to know when the application is finished for now, but I feel that's probably just kicking the can down the road. Any transport feature that's backed by disposable stuff could benefit from observing application completion in order to prevent ODE's.

**Comment by davidfowl** [comment]
Let’s tlsk about this tomorrow. I dunno if it needs to be this complicated but I appreciate that you have given it a lot of thought.

**Comment by halter73** [comment]
I'm not saying those are the only options, I was basically just listing the things I've considered.

File: src/Kestrel.Core/Internal/HttpConnection.cs [hunk 1]
```diff
@@ -105,6 +105,10 @@ public Task StartRequestProcessing<TContext>(IHttpApplication<TContext> applicat
         {
             try
             {
+                // TODO: When we start tracking all connection middleware for shutdown, go back
```
Comment (davidfowl) [inline src/Kestrel.Core/Internal/HttpConnection.cs]: Why did you move this?

Comment (halter73) [reply src/Kestrel.Core/Internal/HttpConnection.cs]: So tests can verify connection processing wrapped up by checking that ConnectionStop was called once WebHost.Dispose() completed. Since the Task returned by this method isn't tracked via the HttpConnectionManager, WebHost.Dispose() can complete before ConnectionStop was logged causing test failures.

The plan is to move it back once we track all connections including those not handled by the HttpMiddleware. It will probably makes sense to move the EventSource logs to here at the same time.

Comment (davidfowl) [reply src/Kestrel.Core/Internal/HttpConnection.cs]: All of these temporary hacks to make the tests pass are beginning to bug me. Does this at least make the tests more green now?

Comment (muratg) [reply src/Kestrel.Core/Internal/HttpConnection.cs]: @davidfowl would you recommend a different approach?

Comment (halter73) [reply src/Kestrel.Core/Internal/HttpConnection.cs]: Personally, I think it will be trivial and obvious to switch back when we move connection tracking downstack. I can create a new issue to track it. My focus atm is making the tests more reliable.

**Comment by halter73** [comment]
#2547

File: src/Kestrel.Transport.Sockets/Internal/SocketConnection.cs [hunk 2]
```diff
@@ -105,16 +105,29 @@ public void Dispose()
...
                 await ProcessReceives();
             }
-            catch (SocketException ex) when (ex.SocketErrorCode == SocketError.ConnectionReset)
+            catch (SocketException ex) when (IsConnectionResetError(ex.SocketErrorCode))
             {
-                error = new ConnectionResetException(ex.Message, ex);
-                _trace.ConnectionReset(ConnectionId);
+                _trace.LogTrace("DoReceive Reset (_aborted? {aborted}) {status}: {ex}", _aborted, (int)ex.SocketErrorCode, ex);
+                // A connection reset can be reported as SocketError.ConnectionAborted on Windows
+                if (!_aborted)
```
Comment (pakrym) [inline src/Kestrel.Transport.Sockets/Internal/SocketConnection.cs]: Merge with when condition when removing logs.

File: test/Kestrel.FunctionalTests/ResponseTests.cs [hunk 3]
```diff
@@ -2485,23 +2519,83 @@ public Task ResponseStatusCodeSetBeforeHttpContextDisposedRequestMalformedReadIg
...
+                {
+                    await connection.Send(
+                        "GET / HTTP/1.1",
+                        "Host:",
+                        "",
+                        "");
 
-                    await appCompletedTcs.Task.TimeoutAfter(TestConstants.DefaultTimeout);
+                    // Force a reset
+                    connection.Socket.LingerState = new LingerOption(true, 0);
```
Comment (pakrym) [inline test/Kestrel.FunctionalTests/ResponseTests.cs]: Make it a TestConnection method.

File: test/Kestrel.FunctionalTests/ResponseTests.cs [hunk 4]
```diff
@@ -2485,23 +2519,83 @@ public Task ResponseStatusCodeSetBeforeHttpContextDisposedRequestMalformedReadIg
...
+
+            var testContext = new TestServiceContext(mockLoggerFactory.Object);
+
+            // There's not guarantee that the app even gets invoked in this test. The connection reset can be observed
+            // as early as accept.
+            using (var server = new TestServer(context => Task.CompletedTask, testContext, listenOptions))
+            {
+                using (var connection = server.CreateConnection())
+                {
+                    await connection.Send(
```
Comment (pakrym) [inline test/Kestrel.FunctionalTests/ResponseTests.cs]: This test is a bit funky already, but what if we run a loop of these? Around 10-100 times?

File: src/Kestrel.Transport.Sockets/Internal/SocketConnection.cs (hunk 2 above)
Comment (halter73) [reply src/Kestrel.Transport.Sockets/Internal/SocketConnection.cs]: We're swallowing the exception if aborted, we don't want it to fall through to the next catch.

File: test/Kestrel.FunctionalTests/ResponseTests.cs [hunk 5]
```diff
@@ -2485,23 +2519,83 @@ public Task ResponseStatusCodeSetBeforeHttpContextDisposedRequestMalformedReadIg
...
+                // On Linux and macOS, the connection close is still sometimes observed as a FIN despite the LingerState.
+                Assert.True(connectionResetLogged || (!TestPlatformHelper.IsWindows && connectionFinLogged), "Connection reset not logged.");
+            }
+
+            mockKestrelTrace.Verify(t => t.ConnectionStop(It.IsAny<string>()), Times.Once());
+            Assert.True(requestAborted, "RequestAborted token didn't fire.");
+            Assert.False(loggedHigherThanDebug, "Logged message with level higher than debug.");
+        }
+
+
```
Comment (pakrym) [inline test/Kestrel.FunctionalTests/ResponseTests.cs]: Nit: Newline

File: test/shared/TaskTimeoutExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    public static class TaskTimeoutExtensions
+    {
+        public static Task<T> DefaultTimeout<T>(this Task<T> task)
+        {
+            return task.DefaultTimeout();
+        }
+
+        public static Task DefaultTimeout(this Task task)
+        {
+            return task.DefaultTimeout();
```
Comment (pakrym) [inline test/shared/TaskTimeoutExtensions.cs]: `13:24:24   Process is terminated due to StackOverflowException.`
